src/io.ts

```diff
@@ -227,30 +217,68 @@ export function getTodoBatches(
 /**
  * Applies todo changes, either adding or removing, based on batches from `getTodoBatches`.
  *
- * @param todoStorageDir - The .lint-todo storage directory.
+ * @param todoStorageFilePath - The .lint-todo storage file path.
  * @param add - Batch of todos to add.
  * @param remove - Batch of todos to remove.
+ * @param options - An object containing write options.
  */
 export function applyTodoChanges(
-  todoStorageDir: string,
-  add: Map<TodoFileHash, TodoDataV2>,
-  remove: Map<TodoFileHash, TodoDataV2>
+  todoStorageFilePath: string,
+  add: Set<TodoData>,
+  remove: Set<TodoData>
 ): void {
-  for (const [fileHash, todoDatum] of add) {
-    const { dir } = posix.parse(fileHash);
+  const ops = buildTodoOperations(add, remove);
+  const release = lockStorageFile(todoStorageFilePath);
 
-    ensureDirSync(posix.join(todoStorageDir, dir));
-    writeJsonSync(posix.join(todoStorageDir, `${fileHash}.json`), todoDatum);
-  }
+  appendFileSync(todoStorageFilePath, ops);
 
-  for (const [fileHash] of remove) {
-    const { dir } = posix.parse(fileHash);
-    const todoDir = posix.join(todoStorageDir, dir);
+  release();
+}
 
-    unlinkSync(posix.join(todoStorageDir, `${fileHash}.json`));
+/**
+ * Compact strategy to leave only add operations in the todo storage file.
+ *
+ * @param operation - The single line operation read from the todo storage file.
+ * @returns True if the line matches an add operation, otherwise false.
+ */
+export const ADD_OPERATIONS_ONLY = (operation: Operation): boolean => /^add.*/.test(operation);
 
-    if (readdirSync(todoDir).length === 0) {
-      rmdirSync(todoDir);
-    }
+/**
+ * Compact strategy to remove all expired operations from the todo storage file.
+ *
+ * @param operation - The single line operation read from the todo storage file.
+ * @returns True if the operation is not expired, otherwise false.
+ */
+export const EXCLUDE_EXPIRED = (operation: Operation): boolean => {
+  const [, todoDatum] = toTodoDatum(operation);
+
+  return !isExpired(todoDatum.errorDate);
+};
+
+/**
+ * Compacts the .lint-todo storage file based on the compact strategy.
+ *
+ * @param baseDir - The base directory that contains the .lint-todo storage directory.
+ * @param compactStrategy - The strategy to use when compacting the storage file. Default: ADD_OPERATIONS_ONLY
+ */
+export function compactTodoStorageFile(
+  baseDir: string,
+  compactStrategy: (operation: Operation) => boolean = ADD_OPERATIONS_ONLY
+): void {
+  const todoStorageFilePath = getTodoStorageFilePath(baseDir);
+
+  const operations = readTodoStorageFile(todoStorageFilePath).filter((operation) =>
+    compactStrategy(operation)
+  );
+
+  writeTodoStorageFile(todoStorageFilePath, operations);
```

Comment:
This whole thing should be wrapped in a lock

Reply:
Done.